common/utils/findandreplace.js

```diff
@@ -0,0 +1,648 @@
+        }
+
+        // Create the replacement node:
+        var newNode = this.getPortionReplacementNode(endPortion, match);
+
+        node.parentNode.insertBefore(newNode, node);
+
+        if (endPortion.endIndexInNode < node.length) {
+          // ?????
+          // Add `after` text node (after the match)
+          followingTextNode = doc.createTextNode(node.data.substring(endPortion.endIndexInNode));
+          node.parentNode.insertBefore(followingTextNode, node);
+        }
+
+        node.parentNode.removeChild(node);
+
+        this.reverts.push(function() {
+          if (precedingTextNode === newNode.previousSibling) {
+            precedingTextNode.parentNode.removeChild(precedingTextNode);
+          }
+          if (followingTextNode === newNode.nextSibling) {
+            followingTextNode.parentNode.removeChild(followingTextNode);
+          }
+          newNode.parentNode.replaceChild(node, newNode);
+        });
+
+        return newNode;
+      } else {
+        // Replace matchStartNode -> [innerMatchNodes...] -> matchEndNode (in that order)
+
+        precedingTextNode = doc.createTextNode(
+          matchStartNode.data.substring(0, startPortion.indexInNode)
+        );
+
+        followingTextNode = doc.createTextNode(
+          matchEndNode.data.substring(endPortion.endIndexInNode)
+        );
+
+        var firstNode = this.getPortionReplacementNode(startPortion, match);
+
+        var innerNodes = [];
+
+        for (var i = 0, l = innerPortions.length; i < l; ++i) {
+          var portion = innerPortions[i];
+          var innerNode = this.getPortionReplacementNode(portion, match);
+          portion.node.parentNode.replaceChild(innerNode, portion.node);
+          this.reverts.push(
+            (function(portion, innerNode) {
+              return function() {
+                innerNode.parentNode.replaceChild(portion.node, innerNode);
+              };
+            })(portion, innerNode)
+          );
+          innerNodes.push(innerNode);
+        }
+
+        var lastNode = this.getPortionReplacementNode(endPortion, match);
+
+        matchStartNode.parentNode.insertBefore(precedingTextNode, matchStartNode);
+        matchStartNode.parentNode.insertBefore(firstNode, matchStartNode);
+        matchStartNode.parentNode.removeChild(matchStartNode);
+
+        matchEndNode.parentNode.insertBefore(lastNode, matchEndNode);
+        matchEndNode.parentNode.insertBefore(followingTextNode, matchEndNode);
+        matchEndNode.parentNode.removeChild(matchEndNode);
+
+        this.reverts.push(function() {
+          precedingTextNode.parentNode.removeChild(precedingTextNode);
+          firstNode.parentNode.replaceChild(matchStartNode, firstNode);
+          followingTextNode.parentNode.removeChild(followingTextNode);
+          lastNode.parentNode.replaceChild(matchEndNode, lastNode);
+        });
+
+        return lastNode;
+      }
+    }
+  };
+
+  return exposed;
+});
```

Comment:
This looks a lot like a module. Should we be pulling in a dependency instead?

Reply:
It's just used for debugging. It helps adds the emoji's to strings. It'll be removed with the rest of the debugging stuff.